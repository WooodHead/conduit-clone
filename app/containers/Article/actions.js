/*
 *
 * Article actions
 *
 */

import {
  DEFAULT_ACTION,
  FETCH_ARTICLE,
  ARTICLE_LOADED,
  GET_ARTICLE_SLUG,
  COMMENTS_LOADED,
  LOAD_ARTICLE_ERROR,
  LOAD_COMMENTS_ERROR,
  DELETING_ARTICLE,
  DELETE_ARTICLE,
  DELETE_ARTICLE_COMPLETED,
  DELETE_ARTICLE_ERROR,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function fetchArticle(slug) {
  return {
    type: FETCH_ARTICLE,
    slug,
  };
}

export function articleLoaded(article) {
  return {
    type: ARTICLE_LOADED,
    article,
  };
}

export function getArticleSlug(slug) {
  return {
    type: GET_ARTICLE_SLUG,
    slug,
  };
}

export function articleLoadingError() {
  return {
    type: LOAD_ARTICLE_ERROR,
  };
}

export function commentsLoaded(comments) {
  return {
    type: COMMENTS_LOADED,
    comments,
  };
}

export function commentsLoadingError() {
  return {
    type: LOAD_COMMENTS_ERROR,
  };
}

export function deletingArticle() {
  return {
    type: DELETING_ARTICLE,
  };
}

export function deleteArticle(slug) {
  return {
    type: DELETE_ARTICLE,
    slug,
  };
}

export function deleteArticleCompleted() {
  return {
    type: DELETE_ARTICLE_COMPLETED,
  };
}

export function deleteArticleError(err) {
  return {
    type: DELETE_ARTICLE_ERROR,
    err,
  };
}
