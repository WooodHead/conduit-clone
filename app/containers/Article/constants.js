/*
 *
 * Article constants
 *
 */

export const DEFAULT_ACTION = 'app/Article/DEFAULT_ACTION';
export const FETCH_ARTICLE = 'app/Article/FETCH_ARTICLE';
export const ARTICLE_LOADED = 'app/Article/ARTICLE_LOADED';
export const LOAD_ARTICLE_ERROR = 'app/Article/LOAD_ARTICLE_ERROR';
