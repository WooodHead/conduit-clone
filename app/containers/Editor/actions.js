/*
 *
 * Editor actions
 *
 */

import {
  ADD_NEW_POST,
  ADDING_NEW_POST,
  ADD_NEW_POST_COMPLETED,
  ADD_NEW_POST_ERROR,
  FETCH_ARTICLE,
  FETCHING_ARTICLE,
  FETCH_ARTICLE_COMPLETED,
  FETCH_ARTICLE_ERROR,
  UPDATE_ARTICLE,
  UPDATING_ARTICLE,
  UPDATE_ARTICLE_COMPLETED,
  UPDATE_ARTICLE_ERROR,
} from './constants';

export function addNewPost(form) {
  return {
    type: ADD_NEW_POST,
    form,
  };
}

export function addingNewPost() {
  return {
    type: ADDING_NEW_POST,
  };
}

export function addNewPostCompleted(article) {
  return {
    type: ADD_NEW_POST_COMPLETED,
    article,
  };
}

export function addNewPostError() {
  return {
    type: ADD_NEW_POST_ERROR,
  };
}

export function fetchArticle(slug) {
  return {
    type: FETCH_ARTICLE,
    slug,
  };
}

export function fetchingArticle() {
  return {
    type: FETCHING_ARTICLE,
  };
}

export function fetchArticleCompleted(article) {
  return {
    type: FETCH_ARTICLE_COMPLETED,
    article,
  };
}

export function fetchArticleError(err) {
  return {
    type: FETCH_ARTICLE_ERROR,
    err,
  };
}

export function updateArticle(slug, article) {
  return {
    type: UPDATE_ARTICLE,
    slug,
    article,
  };
}

export function updatingArticle() {
  return {
    type: UPDATING_ARTICLE,
  };
}

export function updateArticleCompleted(article) {
  return {
    type: UPDATE_ARTICLE_COMPLETED,
    article,
  };
}

export function updateArticleError(err) {
  return {
    type: UPDATE_ARTICLE_ERROR,
    err,
  };
}
