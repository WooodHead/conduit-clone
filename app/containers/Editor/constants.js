/*
 *
 * Editor constants
 *
 */

export const ADD_NEW_POST = 'app/Editor/ADD_NEW_POST';
export const ADDING_NEW_POST = 'app/Editor/ADDING_NEW_POST';
export const ADD_NEW_POST_COMPLETED = 'app/Editor/ADD_NEW_POST_COMPLETED';
export const ADD_NEW_POST_ERROR = 'app/Editor/ADD_NEW_POST_ERROR';
export const FETCH_ARTICLE = 'app/Editor/FETCH_ARTICLE';
export const FETCHING_ARTICLE = 'app/Editor/FETCHING_ARTICLE';
export const FETCH_ARTICLE_COMPLETED = 'app/Editor/FETCH_ARTICLE_COMPLETED';
export const FETCH_ARTICLE_ERROR = 'app/Editor/FETCH_ARTICLE_ERROR';
export const UPDATING_ARTICLE = 'app/Editor/UPDATING_ARTICLE';
export const UPDATE_ARTICLE = 'app/Editor/UPDATE_ARTICLE';
export const UPDATE_ARTICLE_COMPLETED = 'app/Editor/UPDATE_ARTICLE_COMPLETED';
export const UPDATE_ARTICLE_ERROR = 'app/Editor/UPDATE_ARTICLE_ERROR';
