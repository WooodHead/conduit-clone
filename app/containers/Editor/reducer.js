/*
 *
 * Editor reducer
 *
 */

import { fromJS } from 'immutable';
import {
  ADDING_NEW_POST,
  ADD_NEW_POST_COMPLETED,
  ADD_NEW_POST_ERROR,
  FETCHING_ARTICLE,
  FETCH_ARTICLE_COMPLETED,
  FETCH_ARTICLE_ERROR,
  UPDATING_ARTICLE,
  UPDATE_ARTICLE_COMPLETED,
  UPDATE_ARTICLE_ERROR,
} from './constants';

const initialState = fromJS({
  fetching: false,
  updating: false,
  success: false,
  error: false,
  slug: null,
  article: null,
});

function editorReducer(state = initialState, action) {
  switch (action.type) {
    case ADDING_NEW_POST:
      return state.set('fetching', true);
    case ADD_NEW_POST_COMPLETED:
      return state
        .set('fetching', false)
        .set('success', true)
        .set('slug', action.article.slug);
    case ADD_NEW_POST_ERROR:
      return state
        .set('fetching', false)
        .set('error', true);
    case FETCHING_ARTICLE:
      return state.set('fetching', true);
    case FETCH_ARTICLE_COMPLETED:
      return state
        .set('fetching', false)
        .set('error', false)
        .set('article', action.article);
    case FETCH_ARTICLE_ERROR:
      return state
        .set('fetching', false)
        .set('error', true);
    case UPDATING_ARTICLE:
      return state
        .set('updating', true);
    case UPDATE_ARTICLE_COMPLETED:
      return state
        .set('updating', false)
        .set('success', true)
        .set('article', action.article)
        .set('slug', action.article.slug);
    case UPDATE_ARTICLE_ERROR:
      return state
        .set('updating', false)
        .set('error', true);
    default:
      return state;
  }
}

export default editorReducer;
