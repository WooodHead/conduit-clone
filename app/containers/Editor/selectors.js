import { createSelector } from 'reselect';

/**
 * Direct selector to the editor state domain
 */
const selectEditorDomain = (state) => state.get('editor');

/**
 * Default selector used by Settings
 */
const makeSelectFetching = () => createSelector(
  selectEditorDomain,
  (substate) => substate.get('fetching')
);

const makeSelectSuccess = () => createSelector(
  selectEditorDomain,
  (substate) => substate.get('success')
);

const makeSelectSlug = () => createSelector(
  selectEditorDomain,
  (substate) => substate.get('slug')
);

const makeSelectError = () => createSelector(
  selectEditorDomain,
  (substate) => substate.get('error')
);

const makeSelectArticle = () => createSelector(
  selectEditorDomain,
  (substate) => substate.get('article')
);

export {
  selectEditorDomain,
  makeSelectFetching,
  makeSelectSuccess,
  makeSelectSlug,
  makeSelectError,
  makeSelectArticle,
};
