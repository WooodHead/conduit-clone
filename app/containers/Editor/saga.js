import { takeLatest, put, call } from 'redux-saga/effects';
import { ADD_NEW_POST } from 'containers/Editor/constants';

import {
  addingNewPost,
  addNewPostCompleted,
  addNewPostError,
} from 'containers/Editor/actions';
import api from '../../utils/api';

export function* addNewPost(params) {
  yield put(addingNewPost());
  try {
    const payload = yield call(api.Article.add, params.form);
    yield put(addNewPostCompleted(payload.article));
  } catch (err) {
    yield put(addNewPostError(err));
  }
}

export default function* onAddNewPost() {
  yield takeLatest(ADD_NEW_POST, addNewPost);
}
