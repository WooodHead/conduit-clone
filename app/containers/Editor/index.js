/**
 *
 * Editor
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { push } from 'react-router-redux';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { Container, Alert } from 'reactstrap';
import Loader from 'components/Loader';
import PostForm from 'components/PostForm';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import { makeSelectFetching, makeSelectSuccess, makeSelectSlug, makeSelectError, makeSelectArticle, makeSelectUpdating } from './selectors';
import reducer from './reducer';
import saga from './saga';
import {
  addNewPost,
  updateArticle,
  fetchArticle,
  } from './actions';

export class Editor extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor() {
    super();
    this.onSubmit = this.onSubmit.bind(this);
  }

  componentDidMount() {
    /* istanbul ignore next */
    if (this.isValidSlug() && !this.isValidLocationState()) {
      this.props.fetchArticle(this.props.match.params.slug);
    }
  }

  componentWillReceiveProps(nextProps) {
    /* istanbul ignore next */
    if (nextProps.success && nextProps.slug !== null && nextProps.slug !== '' && !nextProps.updating) {
      this.props.redirectToSlug(nextProps.slug, nextProps.article);
    }
  }

  onSubmit(form) {
    if (this.isValidSlug() || this.isValidLocationState()) {
      this.props.updateArticle(this.props.match.params.slug, form);
    } else {
      this.props.addNewPost(form);
    }
  }

  isValidLocationState() {
    /* istanbul ignore next */
    return (this.props.location && this.props.location.state && this.props.location.state.article);
  }

  isValidSlug() {
    /* istanbul ignore next */
    return (this.props.match && this.props.match.params.slug);
  }

  renderForm() {
    const setIntialValues = this.isValidLocationState() ? this.props.location.state.article : this.props.article;
    return (
      <PostForm
        fetching={this.props.fetching || this.props.updating}
        onSubmit={this.onSubmit}
        initialValues={setIntialValues}
      />
    );
  }

  render() {
    let content = <Loader />;
    if (!this.props.error) {
      if (this.props.fetching === false) {
        content = this.renderForm();
      }
    } else {
      content = (
        <Alert color="danger">
          Something went wrong.
        </Alert>
      );
    }

    return (
      <div>
        <Helmet>
          <title>Editor</title>
          <meta name="description" content="Description of Editor" />
        </Helmet>
        <Container>
          {content}
        </Container>
      </div>
    );
  }
}

Editor.defaultProps = {
  fetching: false,
  success: false,
  slug: null,
  error: false,
  article: null,
};

Editor.propTypes = {
  addNewPost: PropTypes.func.isRequired,
  redirectToSlug: PropTypes.func.isRequired,
  fetchArticle: PropTypes.func,
  updateArticle: PropTypes.func,
  fetching: PropTypes.bool.isRequired,
  updating: PropTypes.bool,
  success: PropTypes.bool,
  slug: PropTypes.string,
  error: PropTypes.bool,
  article: PropTypes.object,
  match: PropTypes.shape({
    params: PropTypes.shape({
      slug: PropTypes.string,
    }),
  }),
  location: PropTypes.shape({
    state: PropTypes.shape({
      article: PropTypes.object,
    }),
  }),
};

const mapStateToProps = createStructuredSelector({
  fetching: makeSelectFetching(),
  updating: makeSelectUpdating(),
  success: makeSelectSuccess(),
  error: makeSelectError(),
  slug: makeSelectSlug(),
  article: makeSelectArticle(),
});

export function mapDispatchToProps(dispatch) {
  return {
    addNewPost: (form) => dispatch(addNewPost(form)),
    updateArticle: (slug, article) => dispatch(updateArticle(slug, article)),
    redirectToSlug: (slug) => dispatch(push(`/article/${slug}`)),
    fetchArticle: (slug) => dispatch(fetchArticle(slug)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'editor', reducer });
const withSaga = injectSaga({ key: 'editor', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(Editor);
