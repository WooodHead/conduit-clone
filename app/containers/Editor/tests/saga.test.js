/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeLatest, put } from 'redux-saga/effects';
import { ADD_NEW_POST } from 'containers/Editor/constants';
import { addNewPostError } from '../actions';
import onAddNewPost, { addNewPost } from '../saga';

describe('addNewPost Saga', () => {
  let addNewPostGenerator;
  beforeEach(() => {
    const params = {
      form: {
        foo: 'bar',
      },
    };
    addNewPostGenerator = addNewPost(params);
    let putDescriptor = addNewPostGenerator.next().value;
    expect(putDescriptor).toMatchSnapshot();

    const callDescriptor = addNewPostGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();

    putDescriptor = addNewPostGenerator.next({ article: { foo: 'bar' } }).value;
    expect(putDescriptor).toMatchSnapshot();
  });

  it('should dispatch the addNewPostError action if it requests throws an error', () => {
    const putDescriptor = addNewPostGenerator.throw('Error').value;
    expect(putDescriptor).toEqual(put(addNewPostError('Error')));
  });
});

describe('onAddNewPost Saga', () => {
  const defaultSaga = onAddNewPost();
  it('should start task to watch for ADD_NEW_POST action', () => {
    const takeLatestDescriptor = defaultSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(ADD_NEW_POST, addNewPost));
  });
});
