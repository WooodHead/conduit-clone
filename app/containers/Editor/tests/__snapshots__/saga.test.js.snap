// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addNewPost Saga should dispatch the addNewPostError action if it requests throws an error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/Editor/ADDING_NEW_POST",
    },
    "channel": null,
  },
}
`;

exports[`addNewPost Saga should dispatch the addNewPostError action if it requests throws an error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      Object {
        "foo": "bar",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`addNewPost Saga should dispatch the addNewPostError action if it requests throws an error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "article": Object {
        "foo": "bar",
      },
      "type": "app/Editor/ADD_NEW_POST_COMPLETED",
    },
    "channel": null,
  },
}
`;

exports[`fetchArticle Saga should dispatch the fetchArticleError action if it requests throws an error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/Editor/FETCHING_ARTICLE",
    },
    "channel": null,
  },
}
`;

exports[`fetchArticle Saga should dispatch the fetchArticleError action if it requests throws an error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      undefined,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchArticle Saga should dispatch the fetchArticleError action if it requests throws an error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "article": Object {
        "foo": "bar",
      },
      "type": "app/Editor/FETCH_ARTICLE_COMPLETED",
    },
    "channel": null,
  },
}
`;

exports[`updateArticle Saga should dispatch the updateArticleError action if it requests throws an error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "app/Editor/UPDATING_ARTICLE",
    },
    "channel": null,
  },
}
`;

exports[`updateArticle Saga should dispatch the updateArticleError action if it requests throws an error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "sample-slug",
      Object {
        "foo": "bar",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`updateArticle Saga should dispatch the updateArticleError action if it requests throws an error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "article": Object {
        "foo": "bar",
      },
      "type": "app/Editor/UPDATE_ARTICLE_COMPLETED",
    },
    "channel": null,
  },
}
`;
